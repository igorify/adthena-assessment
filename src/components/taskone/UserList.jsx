import React, { useState } from 'react';
import {
  Row,
  UserInfo,
  Users
} from '@components/styled-components/UserList';
import useFetchUsers from '../../hooks/useFetchUsers';

const setFilter = (setValue) => (e) => {
  const value = e.target && e.target.value;
  setValue(value);
};

function UserList() {
  const [value, setValue] = useState('');
  const data = useFetchUsers(value);

  return (
    <div>
      <div>
        Filter:
        <input
          type="text"
          onChange={setFilter(setValue)}
          value={value}
          placeholder="Enter username"
        />
      </div>
      <Users>
        {!!data.length && data.map((user) => (
          <Row key={user.id} data-testid="userWrapper">
            <UserInfo>
              <span>{`Name: ${user.name}`}</span>
              <span>{`Username: ${user.username}`}</span>
            </UserInfo>
            <div>
              <div>
                <span>{user.address.street}</span>
                <span>{user.address.suite}</span>
                <span>{user.address.city}</span>
                <span>{user.address.zipcode}</span>
              </div>
              <div>
                <span>{user.email}</span>
                <span>{user.phone}</span>
              </div>
            </div>
          </Row>
        ))}
      </Users>
    </div>
  );
}

export default UserList;
