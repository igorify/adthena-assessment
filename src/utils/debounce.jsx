export const DEFAULT_DEBOUNCE_TIMEOUT = 5000;

export default function debounce(func, wait = DEFAULT_DEBOUNCE_TIMEOUT) {
  let timeout = null;

  const cleanup = () => {
    if (timeout) clearTimeout(timeout);
  };

  return (...args) => {
    cleanup();
    timeout = setTimeout(() => func.apply(this, args), wait);
  };
}
