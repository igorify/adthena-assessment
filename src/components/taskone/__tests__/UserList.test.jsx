import React from 'react';
import {
  render, act, screen, fireEvent
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import fetch from 'node-fetch';
import fetchMock from 'fetch-mock';
import UserList from '../UserList';
import usersMock from '../../../mocks/usersMock';
import { DEFAULT_DEBOUNCE_TIMEOUT } from '../../../utils/debounce';

import 'jest-styled-components';

const wait = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

describe('UserList Component', () => {
  beforeAll(() => {
    jest.setTimeout(10000);
    global.fetch = fetch;
  });

  afterEach(() => {
    fetchMock.restore();
  });

  it('should render component correctly', async () => {
    const [user] = usersMock;
    fetchMock.once('https://jsonplaceholder.typicode.com/users', [user]);

    let result;
    await act(async () => {
      result = render(<UserList />);
    });
    expect(result).toMatchSnapshot();
  });

  it('should render component correctly without users data', async () => {
    fetchMock.once('https://jsonplaceholder.typicode.com/users', []);

    let result;
    await act(async () => {
      result = render(<UserList />);
    });
    expect(result).toMatchSnapshot();
  });

  it('should render corrent output of compoment after search update', async () => {
    const [bretUser] = usersMock;
    fetchMock.once('https://jsonplaceholder.typicode.com/users', usersMock);
    fetchMock.once('https://jsonplaceholder.typicode.com/users?username=Bret', [bretUser]);

    await act(async () => {
      render(<UserList />);
    });

    const input = screen.getByPlaceholderText('Enter username');

    const elements = screen.getAllByTestId('userWrapper');
    expect(elements.length).toBe(usersMock.length);
    expect(input.value).toBe('');

    fireEvent.change(input, { target: { value: 'Bret' } });
    await wait(DEFAULT_DEBOUNCE_TIMEOUT);

    const elementAfterUpdate = screen.getAllByTestId('userWrapper');
    expect(elementAfterUpdate.length).toBe(1);
    expect(input.value).toBe('Bret');
  });

  it('should correctly work debounce on input search', async () => {
    const [bretUser] = usersMock;
    fetchMock.once('https://jsonplaceholder.typicode.com/users', usersMock);
    fetchMock.once('https://jsonplaceholder.typicode.com/users?username=Bret', [bretUser]);

    const useEffectSpy = jest.spyOn(React, 'useEffect');

    await act(async () => {
      render(<UserList />);
    });

    const input = screen.getByPlaceholderText('Enter username');
    userEvent.type(input, 'Bret');

    expect(fetchMock.calls().length).toBe(1);
    expect(useEffectSpy).toHaveBeenCalledTimes(6);
  });
});
