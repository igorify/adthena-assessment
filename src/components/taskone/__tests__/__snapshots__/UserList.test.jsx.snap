// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList Component should render component correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-bottom: 10px;
  border: 1px solid #000;
  padding: 20px;
}

.c1 span {
  display: block;
  margin-bottom: 5px;
}

.c2 {
  border-right: 1px solid #000;
  text-align: left;
  width: 260px;
}

.c0 {
  max-height: 300px;
  overflow: scroll;
  margin-top: 15px;
}

<body>
    <div>
      <div>
        <div>
          Filter:
          <input
            placeholder="Enter username"
            type="text"
            value=""
          />
        </div>
        <div
          class="c0"
        >
          <div
            class="c1"
            data-testid="userWrapper"
          >
            <div
              class="c2"
            >
              <span>
                Name: Leanne Graham
              </span>
              <span>
                Username: Bret
              </span>
            </div>
            <div>
              <div>
                <span>
                  Kulas Light
                </span>
                <span>
                  Apt. 556
                </span>
                <span>
                  Gwenborough
                </span>
                <span>
                  92998-3874
                </span>
              </div>
              <div>
                <span>
                  Sincere@april.biz
                </span>
                <span>
                  1-770-736-8031 x56442
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div>
        Filter:
        <input
          placeholder="Enter username"
          type="text"
          value=""
        />
      </div>
      <div
        class="UserList__Users-g96ac2-2 hIhCFv"
      >
        <div
          class="UserList__Row-g96ac2-0 hHBWAY"
          data-testid="userWrapper"
        >
          <div
            class="UserList__UserInfo-g96ac2-1 eGETzj"
          >
            <span>
              Name: Leanne Graham
            </span>
            <span>
              Username: Bret
            </span>
          </div>
          <div>
            <div>
              <span>
                Kulas Light
              </span>
              <span>
                Apt. 556
              </span>
              <span>
                Gwenborough
              </span>
              <span>
                92998-3874
              </span>
            </div>
            <div>
              <span>
                Sincere@april.biz
              </span>
              <span>
                1-770-736-8031 x56442
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`UserList Component should render component correctly without users data 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    .c0 {
  max-height: 300px;
  overflow: scroll;
  margin-top: 15px;
}

<div>
      <div>
        <div>
          Filter:
          <input
            placeholder="Enter username"
            type="text"
            value=""
          />
        </div>
        <div
          class="c0"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div>
        Filter:
        <input
          placeholder="Enter username"
          type="text"
          value=""
        />
      </div>
      <div
        class="UserList__Users-g96ac2-2 hIhCFv"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
